### ISSUE TITLE
Notes related to purchase, upgrade and cancellation flows

### ISSUE THREAD

**Issue by brentv** [yola/p-plesk#9]
_From @brentv on November 20, 2015 17:43_

### Purchase flow and states from within Plesk extension:

1. Customer buys a domain and creates their site under the trial plan. At this stage they can click `manage your site` which takes them to WBB. The status field is set to `active trial`. 
2. Customer returns after trial expires. Now the `manage your site` button changes to `purchase` (which goes to the billing system) and the status field is set to `expired trial`. Recommend messaging to explain that the trial has expired and a purchase is required.
3. Customer purchases a plan. Status is set to `active` and the `manage your site` action is enabled. There is a second option of `manage your subscription`.
4. Customers' subscription is suspended (due to lack of payment or other reason). `manage your site` is disabled and the status is set to `suspended`. The only option is to `manage your subscription` which should allow them to reactivate any and all previous purchases. 
5. Customer cancels their subscription. `manage your site` is disabled and the status is set to `cancelled`. The only option is to `manage your subscription` which should allow them to reactive any and all previous purchases. 

### Subscription and add-on cancellation

1. Customers can cancel subscriptions or add-ons at any time but no prorata refunds will be given.
2. Cancelling the subscription plan requires that all associated add-ons be cancelled as well. The reason for this is that add-ons cannot function without an active subscription plan.

See [related ticket](https://github.com/yola/p-whitelabel/issues/543).

### AdManager purchase flow

1. User clicks `Buy AdManager` inside of WHMCS. 
2. User is redirected to AdManager in WBB. User can now create a campaign or purchase a campaign that has already been configured.
3. User clicks `purchase` or `set budget` (aka upgrade/downgrade) on campaign in AdManager.
4. User is redirected with campaign id parameters to WHMCS.
5. WHMCS ([parameterization to be provided](https://github.com/yola/p-odin/issues/83)):
   1. In the case of purchasing a new campaign, show all available plans for purchase.
   2. In the case of an upgrade/downgrade, show the current plan as disabled and allow the high or lower plans to be purchased.
6. User completes payments and get a success page. No redirect to WBB.

### Online store purchase flow

Largely the same as the AdManager purchase flow except that add-ons are mutually exclusive and there no add-on ids.

Estimate of work for the above = ~1 week.

_Copied from original issue: yola/p-odin#105_

**Comment by brentv** [comment]
_From @konrad-modulesgarden on November 23, 2015 12:23_

While our dev team is preparing a plan of implementation for this part, one thing is missing/unclear:

We've decided to handle upgrade/downgrade of an addon when someone is buying admanager for the same campaign or ecommerce for a whole site. However, WHMCS bills client like he was buying a new addon. It may have sense for the upgrade, but downgrade would look like this:

1. Day 1: Client purchase ecommerce unlimited for $50 and will be paying recurringly $50
2. Day 2: Client wants to downgrade to ecommerce basic, he needs to pay $5 (cost of the addon). After purchase we will cancel unlimited ecommerce so client won't pay another $50 recurringly but $5 for the basic. 

In such case, client paid $55 total in two days and has just a basic package. It may not be that harmful if user upgrades after a while or in the middle/by end of his cycle, but it's really far from perfect.

Please let me know if we stick to this "default WHMCS flow" or not. If it's not clear, we can have a short meeting today so I could explain that.

**Comment by brentv** [comment]
@robertbray, my expectation for the above scenario would be that a user can downgrade at any time but that the plan change would only be applied at the end of the current term and the user would only be charged for the new $5 plan at renewal. 

Konrad, would that be possible with WHMCS? 

I'm available for a call on Tuesday morning PT if that makes sense.

**Comment by brentv** [comment]
_From @robertbray on November 24, 2015 11:42_

The default WHMCS flow is how our system works today - SBS always assumes a new purchase for an new term for any upgrade/downgrade.  

The delayed downgrade at time of renewal (that Brent describes) is a something we are planning to implement and could bring that up in priority of it works better here.  The only question is if the fact the term will be misaligned is a potential problem ie. if the user upgrades half way through on the 15th but it starts on the 1st of the next month.

Also would coding around the WHMCS norm be strange for partners if they always expect WHMCS to work this way?

What is your preferred solution to this @konrad-modulesgarden ?

**Comment by brentv** [comment]
_From @konrad-modulesgarden on November 24, 2015 12:18_

@brentv @robertbray I have explained you the only possible scenario with WHMCS and it's just the default behaviour. Any modification to this standard flow is just not a good idea.

A preferred solution would be that when client pays $55 with my previous sample scenario, he gets unlimited to the end of the billing cycle and then it changes to Basic (after a month from initial purchase).

**Comment by brentv** [comment]
_From @konrad-modulesgarden on November 24, 2015 12:19_

but that would have to be handled on the Yola/Topline side. Please let me know if it is/will be possible so we can start improving the module.

**Comment by brentv** [comment]
_From @robertbray on November 24, 2015 20:14_

Konrad the preferred flow will be possible but is not yet.  Once it is ready we will set it as a platform level parameter so that all WHMCS partners will have the delayed downgrade functionality.  In this way you can code to the current API and we can implement this on our side with no changes to your side.

**Comment by brentv** [comment]
Konrad: Robert and I tested and confirmed that the `Topline > Yola` implementation downgrades the subscription immediately. 

> A preferred solution would be that when client pays $55 with my previous sample scenario, he gets unlimited to the end of the billing cycle and then it changes to Basic (after a month from initial purchase).

Topline will need to add support for this. @robertbray, will capture this requirement into its own ticket.
